pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,664 @@
+		if err != nil && !kerrors.IsNotFound(err) {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if homeClusterWorkspace != nil {
+			// The home workspace for the user exists, so let's return it to the user if he requests to see its definition.
+			// We don't need to check any permission before returning the home workspace definition since,
+			// once it has been created, a home workspace is owned by the user.
+
+			homeWorkspace := &tenancyv1beta1.Workspace{}
+			projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
+			responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+			return
+		}
+
+		// Test if the user has the right to see his Home workspace even though it doesn't exists
+		// => test the get verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+		attributes := homeWorkspaceAuthorizerAttributes(user, "get")
+		decision, reason, err := h.authz.Authorize(ctx, attributes)
+		if err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if decision != authorizer.DecisionAllow {
+			responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+			return
+		}
+
+		// TODO: The way we build this URL here is not fully compatible with sharding long-term:
+		// - short term (what is done now): use genericConfig.ExternalAddress
+		// - medium term: shards will know their name eventually, and with that we can lookup the external URL of the shard
+		// - long-term: create home workspace already on `~` access in order to make scheduling happen and then we know the real external URL.
+		homeWorkspaceURL := &url.URL{Scheme: "https", Host: h.externalHost, Path: homeLogicalClusterName.Path()}
+
+		parent, name := homeLogicalClusterName.Split()
+		homeWorkspace := &tenancyv1beta1.Workspace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        name,
+				ClusterName: parent.String(),
+			},
+			Spec: tenancyv1beta1.WorkspaceSpec{
+				Type: tenancyv1alpha1.ClusterWorkspaceTypeReference{
+					Name: tenancyv1alpha1.ClusterWorkspaceTypeName("Home"),
+					Path: tenancyv1alpha1.RootCluster.String(),
+				},
+			},
+			Status: tenancyv1beta1.WorkspaceStatus{
+				URL: homeWorkspaceURL.String(),
+			},
+		}
+
+		responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+		return
+	}
+
+	needsAutomaticCreation, workspaceType := h.needsAutomaticCreation(lcluster.Name)
+	if !needsAutomaticCreation {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	isHome := workspaceType == "Home"
+	if foundLocally, retryAfterSeconds, err := h.searchForReadyWorkspaceInLocalInformers(lcluster.Name, isHome, user.GetName()); foundLocally {
```

Comment:
Technically since you're making every method here a field, they can all be `nil` ... unclear how to handle that but we might be making life hard for ourselves

Reply:
If you create the handler through `homeWorkspaceHandlerBuilder.build()`, which is the way we use, you will have all the fields of the `homeWorkspaceFeatureLogic` struct filled with the implementation, unless you explicitly reset one field to nil. 

Only some unit tests reset some of the function fields to overridden mocks in order to make cut the dependency between tested functions.

Reply:
see this comment answer for more context: https://github.com/kcp-dev/kcp/pull/1373#discussion_r916303112

Reply:
Assuming non-nil is fine and totally common in Golang. I wished Golang had a non-nil-pointer type, but not in this life.

Reply:
@davidfestal @sttts I can count on many hands the number of times that making such an assumption ends up being incorrect. Prow ends up with panics and we have a service outage. Not fun. Golang doesn't have constructors, @davidfestal , so there's no typesafe way to ensure this.